libopae++/include/opaec++/dma_buffer.h

```diff
@@ -0,0 +1,248 @@
+// Copyright(c) 2017, Intel Corporation
+//
+// Redistribution  and  use  in source  and  binary  forms,  with  or  without
+// modification, are permitted provided that the following conditions are met:
+//
+// * Redistributions of  source code  must retain the  above copyright notice,
+//   this list of conditions and the following disclaimer.
+// * Redistributions in binary form must reproduce the above copyright notice,
+//   this list of conditions and the following disclaimer in the documentation
+//   and/or other materials provided with the distribution.
+// * Neither the name  of Intel Corporation  nor the names of its contributors
+//   may be used to  endorse or promote  products derived  from this  software
+//   without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING,  BUT NOT LIMITED TO,  THE
+// IMPLIED WARRANTIES OF  MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED.  IN NO EVENT  SHALL THE COPYRIGHT OWNER  OR CONTRIBUTORS BE
+// LIABLE  FOR  ANY  DIRECT,  INDIRECT,  INCIDENTAL,  SPECIAL,  EXEMPLARY,  OR
+// CONSEQUENTIAL  DAMAGES  (INCLUDING,  BUT  NOT LIMITED  TO,  PROCUREMENT  OF
+// SUBSTITUTE GOODS OR SERVICES;  LOSS OF USE,  DATA, OR PROFITS;  OR BUSINESS
+// INTERRUPTION)  HOWEVER CAUSED  AND ON ANY THEORY  OF LIABILITY,  WHETHER IN
+// CONTRACT,  STRICT LIABILITY,  OR TORT  (INCLUDING NEGLIGENCE  OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE,  EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+#pragma once
+#include <memory>
+#include <cstdint>
+#include <vector>
+#include <initializer_list>
+#include <chrono>
+#include <thread>
+
+#include <opae/buffer.h>
+#include <opaec++/handle.h>
+
+namespace opae
+{
+namespace fpga
+{
+namespace types
+{
+
+class dma_buffer : public std::enable_shared_from_this<dma_buffer>
+{
+public:
+    typedef std::size_t size_t;
+    typedef std::shared_ptr<dma_buffer> ptr_t;
+
+    /** dma_buffer destructor.
+     */
+    virtual ~dma_buffer();
+
+    /** dma_buffer factory method - allocate a dma_buffer.
+     * @param[in] handle The handle used to allocate the buffer.
+     * @param[in] len    The length in bytes of the requested buffer.
+     * @return A valid dma_buffer smart pointer on success, or an
+     * empty smart pointer on failure.
+     */
+    static dma_buffer::ptr_t allocate(handle::ptr_t handle, size_t len);
```

Comment:
What about using a pre-allocated buffer? Should we add a method called `use_buffer(uint8_t *addr)` 
?

Reply:
Added this as dma_buffer::attach().

Reply:
👍